src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java

```diff
@@ -0,0 +1,203 @@
+
+                bootstrapInjector();
+                final HiveMQServer hiveMQServer = injector.getInstance(HiveMQServer.class);
+                try {
+                    hiveMQServer.start();
+                } catch (final Exception e) {
+                    throw new HiveMQServerException(e);
+                }
+                advanceState(State.RUNNING);
+            }).whenComplete(HiveMQServerException.handler(startFuture));
+        } else {
+            log.warn("Could not start EmbeddedHiveMQ. Reason: \"Already started\"");
+        }
+        return startFuture;
+    }
+
+    private void bootstrapInjector() {
+        if (injector == null) {
+            final HivemqId hiveMQId = new HivemqId();
+            final Injector persistenceInjector =
+                    GuiceBootstrap.persistenceInjector(
+                            systemInformation, metricRegistry, hiveMQId, configurationService);
+
+            try {
+                persistenceInjector.getInstance(PersistenceStartup.class).finish();
+            } catch (final InterruptedException e) {
+                log.error("EmbeddedHiveMQ persistence Startup interrupted.");
+            }
+
+            injector =
+                    GuiceBootstrap.bootstrapInjector(systemInformation, metricRegistry, hiveMQId, configurationService,
+                            persistenceInjector);
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> stop() {
+        if (currentState == State.NOT_STARTED) {
+            log.warn("Could not stop EmbeddedHiveMQ. Reason: \"Not started\"");
+            stopFuture = CompletableFuture.completedFuture(null);
+
+        } else if (currentState.isLower(State.STOPPING)) {
+            advanceState(State.STOPPING);
+            log.info("Stopping EmbeddedHiveMQ.");
+
+            stopFuture = startFuture.thenRunAsync(() -> {
+                final ShutdownHooks shutdownHooks = injector.getInstance(ShutdownHooks.class);
+
+                for (final HiveMQShutdownHook hiveMQShutdownHook : shutdownHooks.getRegistry().values()) {
+                    // We call run, as we want to execute the hooks now, in this thread
+                    //noinspection CallToThreadRun
+                    hiveMQShutdownHook.run();
+                }
+                for (final HiveMQShutdownHook hiveMQShutdownHook : shutdownHooks.getAsyncShutdownHooks()) {
+                    // We call run, as we want to execute the hooks now, in this thread
+                    //noinspection CallToThreadRun
+                    hiveMQShutdownHook.run();
+                }
+
+                shutdownHooks.clearRuntime();
+                metricRegistry.removeMatching(MetricFilter.ALL);
+                injector = null;
+                advanceState(State.STOPPED);
+            });
+        } else {
+            log.warn("Could not stop EmbeddedHiveMQ. Reason: \"Already stopping\"");
```

Comment:
imho this should also return an exceptional future

Reply:
i don't think so as this would break idempotency.

Reply:
Found a new solution for this.